guardrails/embedding.py

```diff
@@ -0,0 +1,87 @@
+from abc import ABC, abstractmethod
+from itertools import islice
+from typing import List, Optional
+
+import numpy as np
+
+from guardrails.llm_providers import openai_embedding_wrapper
+
+
+class EmbeddingBase(ABC):
+    def __init__(
+        self,
+        model: Optional[str],
+        encoding_name: Optional[str],
+        max_tokens: Optional[int],
+    ):
+        self._model = model
+        self._encoding_name = encoding_name
+        self._max_tokens = max_tokens
+
+    @abstractmethod
+    def embed(self, texts: List[str]) -> List[List[float]]:
+        ...
+
+    @abstractmethod
+    def embed_query(self, query: str) -> List[float]:
+        ...
+
+    def len_safe_get_embedding(self, text, average=True) -> List[float]:
+        chunk_embeddings = []
+        chunk_lens = []
+
+        print("encoding ", self._encoding_name)
+        for chunk in EmbeddingBase.chunked_tokens(
+            text=text, encoding_name=self._encoding_name, chunk_length=self._max_tokens
+        ):
+            chunk_embeddings.append(openai_embedding_wrapper(chunk, self._model))
+            chunk_lens.append(len(chunk))
+
+        if average:
+            chunk_embeddings = np.average(chunk_embeddings, axis=0, weights=chunk_lens)
+            chunk_embeddings = chunk_embeddings / np.linalg.norm(
+                chunk_embeddings
+            )  # normalizes length to 1
+        return chunk_embeddings.flatten().tolist()
```

Comment:
Curious about returning a list of floats here instead of an np array

Reply:
We are working with a list of floats across API boundaries so I didn't want to change it. It's ok to use np.array as well, but then we should use that across the board for embedding/vectordb operations. VectorDBs that are remote will require using list floats anyways since the values have to be serialized into json.